File: router-deeplink/code/android/app/src/main/java/com/example/router/domain/usecase/RouteUseCase.kt
package com.example.router.domain.usecase

import com.example.router.domain.model.RouteOutcome
import com.example.router.domain.model.RouteRequest
import com.example.router.domain.port.AuditLogger
import com.example.router.domain.port.Navigator
import com.example.router.domain.port.PreferenceStore
import com.example.router.domain.validation.RouteValidator

class RouteUseCase(
  private val validator: RouteValidator,
  private val navigator: Navigator,
  private val prefs: PreferenceStore,
  private val logger: AuditLogger
) {
  fun handle(request: RouteRequest): RouteOutcome {
    val result = validator.validate(request)
    if (!result.valid) {
      return logger.reject(request, result.reason ?: "invalid")
    }

    if (request.route == "profile") {
      val key = request.params["toggle"]
      val value = request.params["value"]?.toBoolean() ?: false
      if (!key.isNullOrBlank()) {
        prefs.set(key, value)
      }
    }

    navigator.navigate(request.route, request.params)
    return logger.accept(request)
  }
}

File: router-deeplink/code/ios/RouterAppApp/RouterCore/Domain/UseCase/RouteUseCase.swift
import Foundation

final class RouteUseCase {
  private let validator: RouteValidator
  private let navigator: Navigator
  private let prefs: PreferenceStore
  private let logger: AuditLogger

  init(validator: RouteValidator, navigator: Navigator, prefs: PreferenceStore, logger: AuditLogger) {
    self.validator = validator
    self.navigator = navigator
    self.prefs = prefs
    self.logger = logger
  }

  func handle(_ request: RouteRequest) -> RouteOutcome {
    let result = validator.validate(request)
    if !result.valid {
      return logger.reject(request, reason: result.reason ?? "invalid")
    }

    if request.route == "profile" {
      let key = request.params["toggle"] ?? ""
      let value = (request.params["value"] as NSString?)?.boolValue ?? false
      if !key.isEmpty {
        _ = prefs.set(key, value)
      }
    }

    navigator.navigate(route: request.route, params: request.params)
    return logger.accept(request)
  }
}
